### PR TITLE
Use common progressbar function for cat and get

### PR DESCRIPTION
**Pull request by rht**
Where the progressbar wrapper is put before the stream is turned into tar.

### CODE REVIEW

**Comment by jbenet** [comment]
I think this needs rebase.

File: unixfs/archive/tar/writer.go [hunk 1]
```diff
@@ -19,18 +20,29 @@ import (
...
+	TarW           *tar.Writer
+	progressBarOut io.Writer
+	progressBar    *progressbar.ProgressBar
 
 	ctx cxt.Context
 }
 
 // NewWriter wraps given io.Writer.
-func NewWriter(ctx cxt.Context, dag mdag.DAGService, archive bool, compression int, w io.Writer) (*Writer, error) {
+func NewWriter(ctx cxt.Context, dag mdag.DAGService, archive bool, compression int, w io.Writer, progressBarOut io.Writer, size uint64) (*Writer, error) {
```
Comment (jbenet) [inline unixfs/archive/tar/writer.go]: No, this is not the right way to do this. the prograss bar stuff does not belong here. All this stuff is already complicated, and this makes it way worse. there must be _some_ way to report out the information from the Writer.

The simplicity and cleanliness of Interfaces really matters to their usability.

Comment (jbenet) [inline unixfs/archive/tar/writer.go]: and their maintenance. I know the codebase is not great in this regard, but we should always strive to make it better.

Comment (rht) [inline unixfs/archive/tar/writer.go]: How about adding a channel telling the size being read to DAGService? Though a callback is more generic to allow outputting `Object` as in the requirement in #1778.

**Comment by jbenet** [comment]
@rht let's walk through the problem again. is the issue that the progress cannot be estimated because the output is a tar archive, and tar adds a bunch of overhead? So the proposed solution here is to measure the bytes _before_ they're tarred?

--- 

some other possible solutions:
1. estimate the overhead, by counting the objects, estimating an upper bound based on the size of tar headers and possilbe zero padding, overshoot, and use that as a 100%. 
  - we should be able to get within 10% of accuracy, so it's fine if 90%-100% is immediate in the progress bar. 
  - this way, the interfaces remain unchanged, and the progress bar mess is kept contained
2.  (As you say) the writer could emit objects written in a sideband channel. this is similar however to just Tee-ing the (unarchived) writer, and using one output for the archive, and the other for the progress bar. This would be cleaner.
  - one possibility is to break apart the DagArchive writer.

- I suspect (1) is way easier.
- I suspect breaking apart the DagArchive into two functions (one to with Archives, one not) will make (2) easier.
- ideally, _all_ the progress bar stuff could happen entirely client-side, not on the daemon. Meaning, if we get the daemon to report total size, or even report progress updates at given intervals, the actual progress bar stuff can stay in the client side).

**Comment by rht** [comment]
> estimate the overhead, by counting the objects, estimating an upper bound based on the size of tar headers and possilbe zero padding, overshoot, and use that as a 100%.

Yes it is possible to do this accurately for tarred object, it's just that the entire dag has to be traversed.

On second thought, the progressbar should only report progress piped to the client (the tar.gz-ed stream) that it doesn't make sense to tell the pre-tar.gz stream.

**Comment by rht** [comment]
(there is the client, and then there is the local node...)

**Comment by jbenet** [comment]
- Yes it is possible to do this accurately for tarred object, it's just that the entire dag has to be traversed.

hmm yes. forgot dags don't include the total # of objects underneath. :/

**Comment by rht** [comment]
Just removed the progressbar on dagreader.
Now this PR contains only a refactor.

**Comment by rht** [comment]
Either:
1. daemon reports to client its download progress via channel, which is not necessarily the stream piped to the client
2. daemon knows the tar.gz / .gz file size before hand; progress count happens entirely client side.

File: core/commands/get.go [hunk 2]
```diff
@@ -159,8 +181,7 @@ func (gw *getWriter) writeArchive(r io.Reader, fpath string) error {
 	defer file.Close()
 
 	fmt.Fprintf(gw.Out, "Saving archive to %s\n", fpath)
-	bar, barR := progressBarForReader(gw.Err, r)
-	bar.Start()
```
Comment (jbenet) [inline core/commands/get.go]: we should keep the Start with the Finish so it's symmetric calls. less error prone. think of these three lines like:

```go
lk := sync.Lock{}
lk.Lock()
defer lk.Unlock()
```

**Comment by jbenet** [comment]
:+1: to the refactor. though:
- [ ] keep the `Start` call symmetric to `Finish` (i.e. not hidden)

**Pull request merged**